Mathlib/Data/Finset/Antidiagonal.lean

```diff
@@ -0,0 +1,195 @@
+/-
+Copyright (c) 2023 Antoine Chambert-Loir and María Inés de Frutos-Fernández. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Antoine Chambert-Loir, María Inés de Frutos-Fernández, Bhavik Mehta
+-/
+
+import Mathlib.Data.Finset.Basic
+import Mathlib.Data.Finsupp.Defs
+import Mathlib.Data.Finsupp.Interval
+import Mathlib.Algebra.Order.Sub.Defs
+
+/-! # Antidiagonal with values in general types
+
+We define a type class `Finset.HasAntidiagonal A` which contains a function
+`antidiagonal : A → Finset (A × A)` such that `antidiagonal n`
+is the Finset of all pairs adding to `n`, as witnessed by `mem_antidiagonal`.
+
+When `A` is a canonically ordered add monoid with locally finite order
+(for example, taking `A` to be `ℕ`, more generally or `σ →₀ ℕ`),
+this typeclass can be instantiated with `Finset.antidiagonalOfLocallyFinite`.
+This definition applies more generally to `ι →₀ A`
+under the additional assumption `OrderedSub A` that make it a canonically ordered add monoid.
+In fact, we just need an `AddMonoid` with a compatible order,
+finite `Iic`, such that if `a + b = n`, then `a, b ≤ n`,
+and any finiteness condition would be OK.
```

Comment:
I think there's some duplication in this paragraph.

Reply:
I changed it slightly.